File: mercury/latex.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: latex.m
% Main author: nyc
%
% LaTeX rendering of radical expressions.
%
%---------------------------------------------------------------------------%

:- module latex.
:- interface.

:- import_module rad_expr.
:- import_module rational.

%---------------------------------------------------------------------------%

    % Render a radical expression as a LaTeX math-mode string.
    %
:- func latex(rad_expr(rational)) = string.

    % Render with precedence context.
    %
:- func latex_prec(int, rad_expr(rational)) = string.

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- import_module bool.
:- use_module int.
:- import_module integer.
:- import_module list.
:- use_module string.

%---------------------------------------------------------------------------%
% Precedence levels
%---------------------------------------------------------------------------%

:- func prec_add = int.
prec_add = 1.

:- func prec_mul = int.
prec_mul = 2.

:- func prec_neg = int.
prec_neg = 3.

:- func prec_pow = int.
prec_pow = 4.

%---------------------------------------------------------------------------%
% Main rendering
%---------------------------------------------------------------------------%

latex(Expr) = latex_prec(0, Expr).

latex_prec(P, Expr) = Result :-
    ( if Expr = re_lit(R) then
        Result = latex_rat(R)
    else if Expr = re_neg(Inner) then
        ( if Inner = re_add(_, _) then
            Terms = flatten_add(Inner),
            NegTerms = list.map(neg_term, Terms),
            Result = parens_if(gt(P, prec_add),
                render_terms(NegTerms))
        else if Inner = re_mul(re_lit(C), Rest) then
            Result = latex_prec(P,
                re_mul(re_lit(rational.'-'(rational.zero, C)), Rest))
        else if Inner = re_lit(R) then
            Result = latex_prec(P, re_lit(rational.'-'(rational.zero, R)))
        else
            Result = parens_if(gt(P, prec_neg),
                string.append("-", latex_prec(prec_neg, Inner)))
        )
    else if Expr = re_add(_, _) then
        Terms = flatten_add(Expr),
        Result = parens_if(gt(P, prec_add), render_terms(Terms))
    else if Expr = re_mul(A, re_inv(B)) then
        Result = parens_if(gt(P, prec_mul),
            string.append_list(["\\frac{", latex_prec(0, A), "}{",
                latex_prec(0, B), "}"]))
    else if Expr = re_mul(re_inv(A), B) then
        Result = parens_if(gt(P, prec_mul),
            string.append_list(["\\frac{", latex_prec(0, B), "}{",
                latex_prec(0, A), "}"]))
    else if Expr = re_inv(Inner2) then
        Result = parens_if(gt(P, prec_mul),
            string.append_list(["\\frac{1}{", latex_prec(0, Inner2), "}"]))
    else if Expr = re_root(2, re_lit(R)),
            rational.'<'(R, rational.zero),
            rational.'+'(R, rational.one) = rational.zero then
        Result = "\\mathrm{i}"
    else if Expr = re_root(2, Inner3) then
        Result = string.append_list(["\\sqrt{",
            latex_radicand(Inner3), "}"])
    else if Expr = re_root(N, Inner4) then
        Result = string.append_list(["\\sqrt[",
            string.int_to_string(N), "]{", latex_radicand(Inner4), "}"])
    else if Expr = re_pow(_, 0) then
        Result = "1"
    else if Expr = re_pow(Base, Exp) then
        ( if int.'<'(Exp, 0) then
            Result = latex_prec(P,
                re_inv(re_pow(Base, int.'-'(0, Exp))))
        else if Exp = 1 then
            Result = latex_prec(P, Base)
        else
            Result = parens_if(gt(P, prec_pow),
                string.append_list([latex_base(Base), "^{",
                    string.int_to_string(Exp), "}"]))
        )
    else if Expr = re_mul(_, _) then
        Factors = flatten_mul(Expr),
        Result = parens_if(gt(P, prec_mul),
            render_factors(Factors))
    else
        Result = "?"
    ).

%---------------------------------------------------------------------------%
% Base rendering for powers
%---------------------------------------------------------------------------%

:- func latex_base(rad_expr(rational)) = string.

latex_base(E) = Result :-
    ( if E = re_root(_, _) then
        Result = string.append_list(["\\left(", latex_prec(0, E), "\\right)"])
    else if E = re_add(_, _) then
        Result = string.append_list(["\\left(", latex_prec(0, E), "\\right)"])
    else if E = re_mul(_, _) then
        Result = string.append_list(["\\left(", latex_prec(0, E), "\\right)"])
    else if E = re_neg(_) then
        Result = string.append_list(["\\left(", latex_prec(0, E), "\\right)"])
    else if E = re_inv(_) then
        Result = string.append_list(["\\left(", latex_prec(0, E), "\\right)"])
    else
        Result = latex_prec(prec_pow, E)
    ).

%---------------------------------------------------------------------------%
% Radicand rendering
%---------------------------------------------------------------------------%

:- func latex_radicand(rad_expr(rational)) = string.

latex_radicand(E) = Result :-
    ( if E = re_lit(R) then
        N = numer(R),
        D = denom(R),
        ( if D = integer.one then
            Result = integer.to_string(N)
        else if integer.'<'(N, integer.zero) then
            Result = string.append_list(["-\\frac{",
                integer.to_string(integer.abs(N)), "}{",
                integer.to_string(D), "}"])
        else
            Result = string.append_list(["\\frac{",
                integer.to_string(N), "}{",
                integer.to_string(D), "}"])
        )
    else
        Result = latex_prec(0, E)
    ).

%---------------------------------------------------------------------------%
% Rational rendering
%---------------------------------------------------------------------------%

:- func latex_rat(rational) = string.

latex_rat(R) = Result :-
    N = numer(R),
    D = denom(R),
    ( if D = integer.one then
        Result = integer.to_string(N)
    else if integer.'<'(N, integer.zero) then
        Result = string.append_list(["-\\frac{",
            integer.to_string(integer.abs(N)), "}{",
            integer.to_string(D), "}"])
    else
        Result = string.append_list(["\\frac{",
            integer.to_string(N), "}{",
            integer.to_string(D), "}"])
    ).

%---------------------------------------------------------------------------%
% Helpers
%---------------------------------------------------------------------------%

:- func gt(int, int) = bool.

gt(A, B) = ( if int.'>'(A, B) then yes else no ).

:- func parens_if(bool, string) = string.

parens_if(yes, S) = string.append_list(["\\left(", S, "\\right)"]).
parens_if(no, S) = S.

:- func neg_term({bool, rad_expr(rational)}) = {bool, rad_expr(rational)}.

neg_term({S, T}) = {flip_bool(S), T}.

:- func flip_bool(bool) = bool.

flip_bool(yes) = no.
flip_bool(no) = yes.

%---------------------------------------------------------------------------%
% Flatten and render add chains
%---------------------------------------------------------------------------%

:- func flatten_add(rad_expr(rational)) = list({bool, rad_expr(rational)}).

flatten_add(Expr) = Result :-
    ( if Expr = re_add(A, B) then
        Result = list.append(flatten_add(A), flatten_add(B))
    else if Expr = re_neg(Inner) then
        Result = list.map(neg_term, flatten_add(Inner))
    else if Expr = re_lit(R), rational.'<'(R, rational.zero) then
        Result = [{no, re_lit(rational.'-'(rational.zero, R))}]
    else if Expr = re_mul(re_neg(A2), B2) then
        Result = [{no, re_mul(A2, B2)}]
    else if Expr = re_mul(_, _) then
        Factors = flatten_mul(Expr),
        ( if Factors = [re_lit(R2) | Rest], rational.'<'(R2, rational.zero) then
            Result = [{no, rebuild_mul([re_lit(rational.'-'(rational.zero, R2)) | Rest])}]
        else
            Result = [{yes, Expr}]
        )
    else
        Result = [{yes, Expr}]
    ).

:- func rebuild_mul(list(rad_expr(rational))) = rad_expr(rational).

rebuild_mul(Fs) = Result :-
    ( if Fs = [] then
        Result = re_lit(rational.one)
    else if Fs = [X] then
        Result = X
    else if Fs = [X | Xs] then
        Result = list.foldl(func(B, Acc) = re_mul(Acc, B), Xs, X)
    else
        Result = re_lit(rational.one)
    ).

:- func render_terms(list({bool, rad_expr(rational)})) = string.

render_terms(Terms) = Result :-
    ( if Terms = [] then
        Result = "0"
    else if Terms = [{S, T} | Rest] then
        Hd = ( if S = yes then latex_prec(prec_add, T)
               else string.append("-", latex_prec(prec_mul, T)) ),
        Result = string.append(Hd, render_rest(Rest))
    else
        Result = "0"
    ).

:- func render_rest(list({bool, rad_expr(rational)})) = string.

render_rest([]) = "".
render_rest([{S, E} | Rest]) = Result :-
    This = ( if S = yes then string.append(" + ", latex_prec(prec_add, E))
             else string.append(" - ", latex_prec(prec_mul, E)) ),
    Result = string.append(This, render_rest(Rest)).

%---------------------------------------------------------------------------%
% Flatten and render mul chains
%---------------------------------------------------------------------------%

:- func flatten_mul(rad_expr(rational)) = list(rad_expr(rational)).

flatten_mul(Expr) = Result :-
    ( if Expr = re_mul(A, B) then
        Result = list.append(flatten_mul(A), flatten_mul(B))
    else
        Result = [Expr]
    ).

:- func render_factors(list(rad_expr(rational))) = string.

render_factors(Factors) = Result :-
    ( if Factors = [] then
        Result = "1"
    else if Factors = [X] then
        Result = latex_prec(prec_mul, X)
    else if Factors = [re_lit(C) | Rest] then
        ( if C = rational.one then
            Result = join_mul(list.map(latex_prec_pow, Rest))
        else if C = rational.'-'(rational.zero, rational.one) then
            Result = string.append("-",
                join_mul(list.map(latex_prec_pow, Rest)))
        else
            Result = string.append_list([latex_rat(C), " \\cdot ",
                join_mul(list.map(latex_prec_pow, Rest))])
        )
    else
        Result = join_mul(list.map(latex_prec_pow, Factors))
    ).

:- func latex_prec_pow(rad_expr(rational)) = string.
latex_prec_pow(E) = latex_prec(prec_pow, E).

:- func join_mul(list(string)) = string.

join_mul([]) = "".
join_mul([X]) = X.
join_mul([X | Xs]) = Result :-
    Xs = [_ | _],
    Result = string.append(X, join_mul_rest(Xs)).

:- func join_mul_rest(list(string)) = string.

join_mul_rest([]) = "".
join_mul_rest([X | Xs]) =
    string.append(string.append(" \\cdot ", X), join_mul_rest(Xs)).

%---------------------------------------------------------------------------%
:- end_module latex.
%---------------------------------------------------------------------------%

File: ocaml/lib/latex.ml
(** LaTeX rendering of radical expressions. *)

open Rad_expr
module R = Rational

(** {2 Precedence levels} *)

let prec_add = 1
let prec_mul = 2
let prec_neg = 3
let prec_pow = 4

(** {2 Helpers} *)

let parens_if cond s =
  if cond then "\\left(" ^ s ^ "\\right)" else s

let latex_rat r =
  let n = R.num r in
  let d = R.den r in
  if Z.equal d Z.one then Z.to_string n
  else if Z.sign n < 0 then
    "-\\frac{" ^ Z.to_string (Z.abs n) ^ "}{" ^ Z.to_string d ^ "}"
  else
    "\\frac{" ^ Z.to_string n ^ "}{" ^ Z.to_string d ^ "}"

let join_mul = function
  | [] -> ""
  | [x] -> x
  | x :: xs -> x ^ String.concat "" (List.map (fun s -> " \\cdot " ^ s) xs)

(** {2 Flattening helpers} *)

let rec flatten_add = function
  | Add (a, b) -> flatten_add a @ flatten_add b
  | Neg e -> List.map (fun (s, t) -> (not s, t)) (flatten_add e)
  | Lit r when R.is_neg r -> [(false, Lit (R.neg r))]
  | Mul (Neg a, b) -> [(false, Mul (a, b))]
  | e -> [(true, e)]

let rec flatten_mul = function
  | Mul (a, b) -> flatten_mul a @ flatten_mul b
  | e -> [e]

(** {2 Rendering} *)

(** Render the base of a power expression with appropriate grouping. *)
let latex_base pp_fn = function
  | Root _ | Add _ | Mul _ | Neg _ | Inv _ as e ->
    "\\left(" ^ pp_fn 0 e ^ "\\right)"
  | e -> pp_fn prec_pow e

(** Render a radicand (inside \\sqrt{...}). *)
let latex_radicand pp_fn = function
  | Lit r ->
    let n = R.num r in
    let d = R.den r in
    if Z.equal d Z.one then Z.to_string n
    else if Z.sign n < 0 then
      "-\\frac{" ^ Z.to_string (Z.abs n) ^ "}{" ^ Z.to_string d ^ "}"
    else
      "\\frac{" ^ Z.to_string n ^ "}{" ^ Z.to_string d ^ "}"
  | e -> pp_fn 0 e

let render_terms pp_fn terms =
  let latex_mul_or_atom = pp_fn prec_mul in
  match terms with
  | [] -> "0"
  | (s, t) :: rest ->
    let hd = if s then pp_fn prec_add t else "-" ^ latex_mul_or_atom t in
    hd ^ String.concat "" (List.map (fun (sign, e) ->
      if sign then " + " ^ pp_fn prec_add e
      else " - " ^ latex_mul_or_atom e) rest)

let render_factors pp_fn = function
  | [] -> "1"
  | [x] -> pp_fn prec_mul x
  | Lit c :: rest when R.is_one c ->
    join_mul (List.map (pp_fn prec_pow) rest)
  | Lit c :: rest when R.equal c R.minus_one ->
    "-" ^ join_mul (List.map (pp_fn prec_pow) rest)
  | Lit c :: rest ->
    latex_rat c ^ " \\cdot " ^ join_mul (List.map (pp_fn prec_pow) rest)
  | fs ->
    join_mul (List.map (pp_fn prec_pow) fs)

(** Render a radical expression as a LaTeX math-mode string. *)
let rec latex_prec p = function
  | Lit r -> latex_rat r
  | Neg e -> begin match e with
    | Add _ ->
      parens_if (p > prec_add)
        (render_terms latex_prec
          (List.map (fun (s, t) -> (not s, t)) (flatten_add e)))
    | Mul (Lit c, rest) ->
      latex_prec p (Mul (Lit (R.neg c), rest))
    | Lit r ->
      latex_prec p (Lit (R.neg r))
    | _ ->
      parens_if (p > prec_neg) ("-" ^ latex_prec prec_neg e)
    end
  | Add _ as e ->
    parens_if (p > prec_add) (render_terms latex_prec (flatten_add e))
  | Mul (a, Inv b) ->
    parens_if (p > prec_mul)
      ("\\frac{" ^ latex_prec 0 a ^ "}{" ^ latex_prec 0 b ^ "}")
  | Mul (Inv a, b) ->
    parens_if (p > prec_mul)
      ("\\frac{" ^ latex_prec 0 b ^ "}{" ^ latex_prec 0 a ^ "}")
  | Inv e ->
    parens_if (p > prec_mul)
      ("\\frac{1}{" ^ latex_prec 0 e ^ "}")
  | Root (2, Lit r) when R.equal r R.minus_one -> "\\mathrm{i}"
  | Root (2, e) -> "\\sqrt{" ^ latex_radicand latex_prec e ^ "}"
  | Root (n, e) ->
    "\\sqrt[" ^ string_of_int n ^ "]{" ^ latex_radicand latex_prec e ^ "}"
  | Pow (_, 0) -> "1"
  | Pow (e, n) when n < 0 ->
    latex_prec p (Inv (Pow (e, -n)))
  | Pow (e, 1) -> latex_prec p e
  | Pow (e, n) ->
    parens_if (p > prec_pow)
      (latex_base latex_prec e ^ "^{" ^ string_of_int n ^ "}")
  | Mul _ as e ->
    parens_if (p > prec_mul) (render_factors latex_prec (flatten_mul e))

(** Render a radical expression as a LaTeX math-mode string. *)
let latex e = latex_prec 0 e
